nxp: linker: Add cmake linker generator support

zejiang0jason opened the pull request:
Support cmake linker generator for some NXP boards, use build option `-DCONFIG_CMAKE_LINKER_GENERATOR=y` to enable the feature

RT500/RT600/RT700/MCXE31 not supported yet, because the these SOCs use code relocation feature, which need to improve cmake linker generator common part.

Depends on: https://github.com/zephyrproject-rtos/zephyr/pull/101039

RobinKastberg committed:
nxp: linker: update kinetis cmake linker script

This is needed to place the `kinetis_flash_config` sections
when using a cmake-based linker script generator.

Signed-off-by: Robin Kastberg <robin.kastberg@iar.com>

zejiang0jason committed:
cmake: linker: arm: Add symbol _flash_used for linker generator

Add _flash_used, it is same with ld method:
```
_flash_used = LOADADDR(.last_section) + SIZEOF(.last_section)
    - __rom_region_start;
```

Remove the one defined in Azephyr/cmake/linker/iar/config_file_script.cmake
in case for conflict

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

zejiang0jason committed:
cmake: linker: arm: Improve VECTOR_ALIGN value

Current `VECTOR_ALIGN` is calculated based on `CONFIG_NUM_IRQS`.
Actually the alignment might be not suitable for the chip.

Change to use chip defined `CONFIG_ARCH_IRQ_VECTOR_TABLE_ALIGN`
first if it is available.

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

zejiang0jason committed:
cmake: linker: arm: Use arch-specific minimum vector alignment

The minimum vector table alignment depends on the ARM architecture
variant due to different VTOR register reserved bits:
- ARMv6-M/ARMv8-M Baseline: bits 0:7 reserved, 256-byte alignment
- ARMv7-M/ARMv8-M Mainline: bits 0:6 reserved, 128-byte alignment

Replace the hardcoded 128-byte minimum with architecture-specific
values that match the logic in arch/arm/core/vector_table.ld.

This ensures correct vector table alignment for all ARM Cortex-M
variants when using the CMake linker generator.

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

zejiang0jason committed:
modules: hal_nxp: linker: Support cmake linker generator

Use cmake linker generator function `zephyr_linker_section_configure`
to place the `DataQuickAccess`, `CodeQuickAccess`, `NonCacheable`,
and `NonCacheable.init` sections in desired memory region.

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

zejiang0jason committed:
soc: nxp: Support cmake linker generator

Update NXP SoC CMakeLists.txt files to use cmake linker generator
functions.

Changes include:
- RW6xx: Configure boot header sections (.flash_conf, .boot_hdr.ivt),
  firmware sections (.fw_cpu1, .fw_cpu2), and SMU sections for
  inter-CPU communication
- LPC55xxx: Configure USB BDT and USB global sections in dedicated
  USB_SRAM
- i.MX RT: Configure boot header and flash config sections with
  proper offsets, separate XMCD data section configuration
- MCXW7xx: Configure RPMsg shared memory section for NBU
- MCXC/MCXE24x/S32K1: Configure flash config section

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

ZhaoxiangJin commented:
> You have added linker sections, and nothing has been deleted/moved. Just wondering, how did it work before?

Previously, we focused on Zephyr SDK, so relying on templates + linker snippets (.ld) was fine. Now, to extend support for IAR/ARM Clang, which don't recognize .ld files, we need to add CMake linker generator support.

ZhaoxiangJin commented:
> > > You have added linker sections, and nothing has been deleted/moved. Just wondering, how did it work before?
> > 
> > 
> > Previously, we focused on Zephyr SDK, so relying on templates + linker snippets (.ld) was fine. Now, to extend support for IAR/ARM Clang, which don't recognize .ld files, we need to add CMake linker generator support.
> 
> So this is a workaround for IAR/Clang limitations. Should we report this to IAR? I had thought that IAR was adapting Zephyr, and we could continue to rely on common templates/snippets/code.

It's hard to say this is a limitation of IAR/Clang, because each toolchain has its own linker format. If we want our code to be compiled correctly by these toolchains, we should use the functions provided by the CMake linker generator to properly place these sections. IAR has already document this. https://docs.zephyrproject.org/latest/develop/toolchains/iar_arm_toolchain.html

hakehuang commented:
I will do a regression on all NXP platforms with IAR compiler after https://github.com/zephyrproject-rtos/zephyr/pull/101039 get merged

butok commented:
> It's hard to say this is a limitation of IAR/Clang, because each toolchain has its own linker format. If we want our code to be compiled correctly by these toolchains, we should use the functions provided by the CMake linker generator to properly place these sections. IAR has already document this. https://docs.zephyrproject.org/latest/develop/toolchains/iar_arm_toolchain.html

Hi @RobinKastberg 

There is the known IAR limitation:
`"The IAR Toolchain uses ilink for linking and depends on Zephyr’s CMAKE_LINKER_GENERATOR. ilink is incompatible with Zephyr’s linker script template, which works with GNU ld."`

Is IAR planning to solve it? 
Or is this something that Zephyr projects must adjust in its own way for each platform?

zejiang0jason commented:
rebase to include latest change in pr: https://github.com/zephyrproject-rtos/zephyr/pull/101039

RobinKastberg commented on soc/nxp/rw/CMakeLists.txt:
```diff
@@ -18,11 +18,108 @@ zephyr_sources_ifdef(CONFIG_PM
+dt_reg_size(SMU2_SIZE PATH ${smu2_data_node})
+
+zephyr_linker_section(
+  NAME .SMU1
```
This is probably not intended, and has little to do with the "." I think. 
Since Regions from device tree generate a default block, and your output section defines a block with the same name, there is a collision.
I will have to take a closer look how to resolve this. But my first reaction is that you can probably change the output section name to work around this. Annoying but should be harmless unless there are tools consuming the output section info.

tejlmand reviewed (commented):
LGTM, didn't spot anything suspicious code-wise, but the section naming collision must be solved before giving a +1

zejiang0jason commented:
convert to draft because the PR depends on is not merged yet: https://github.com/zephyrproject-rtos/zephyr/pull/101039.

zejiang0jason committed:
soc: nxp: rt11xx: Add soc_reset_hook for IAR

Originally the symbol soc_reset_hook in asm is unavailable for IAR.
Add C function soc_reset_hook for IAR.

Signed-off-by: Jason Yu <zejiang.yu@nxp.com>

tejlmand commented:
> Hi @tejlmand , this PR depends on #101039, the non-NXP specific code changes are in that PR, please help review to see whether it can be merged first, thanks.

which again seems to have spawned https://github.com/zephyrproject-rtos/zephyr/pull/102472 that needs merging first.
Guess we should get the cleanup handled first :slightly_smiling_face:

zejiang0jason commented:
Because there are many platforms to support, I plan to use split PRs to update family by family.

> @zejiang0jason there are more platforms need work see below summary

butok commented:
https://github.com/zephyrproject-rtos/zephyr/pull/101039 it depends on has been closed.

hakehuang commented:
@zejiang0jason can you rebase ?
